jena-arq/src/main/java/org/apache/jena/sparql/engine/main/JoinClassifier.java

```diff
@@ -235,4 +243,85 @@ private static boolean safeModifier(Op op) {
             return false ;
         return op instanceof OpDistinct || op instanceof OpReduced || op instanceof OpProject || op instanceof OpList ;
     }
+
+    /** Enumeration of the primitive op types in the SPARQL algebra. */
+    private enum Basis {
+        PATTERN,
+        TABLE,
+        PFUNCTION
+    }
+
+    /**
+     * This method checks whether the given op is based on {@link OpTable} or {@link OpPropFunc}.
+     * If neither is the case then the result is {@link Basis#PATTERN}.
+     * <p>
+     * This method is called for each side of a join from {@link #isLinear(Op, Op)}.
+     * If that side of the join is a property function then {@link Basis#PFUNCTION} is returned.
+     *
+     * <p>
+     * The special handling of property functions is due to that
+     * OpJoin(TABLE, PFUNCTION) needs to be linearized to OpSequence(TABLE, PFUNCTION).
+     *
+     * <p>
+     * This method resolves OpExt to its effective op.
+     */
+    private static Basis getBasis(Op op) {
```

Comment:
This will look better when switch/case/instanceof is available! (Java21, not Java17)